picard/browser/browser.py

```diff
@@ -25,115 +25,169 @@
 
     def stop(self):
-        if self.port > 0:
-            log.debug("Stopping the browser integration")
-            self.port = 0
-            self.tagger.listen_port_changed.emit(self.port)
-            self.close()
+        if self.server:
+            log.info("Stopping the browser integration")
+            self.server.shutdown()
+            self.server.server_close()
+            self.server = None
+            self.listen_port_changed.emit(self.port)
         else:
             log.debug("Browser integration inactive, no need to stop")
 
-    def _process_request(self):
-        conn = self.sender()
-        rawline = conn.readLine().data()
-        log.debug("Browser integration request: %r", rawline)
 
-        def parse_line(line):
-            orig_line = line
-            try:
-                line = line.split()
-                if line[0] == "GET" and "?" in line[1]:
-                    parsed = urlparse(line[1])
-                    args = parse_qs(parsed.query)
-                    if 'id' in args and args['id']:
-                        mbid = args['id'][0]
-                        if not mbid_validate(mbid):
-                            log.error("Browser integration failed: bad mbid %r", mbid)
-                            return False
-
-                        def load_it(loader):
-                            self.tagger.bring_tagger_front()
-                            loader(mbid)
-                            return True
-                        action = parsed.path
-                        if action == '/openalbum':
-                            return load_it(self.tagger.load_album)
-                        elif action == '/opennat':
-                            return load_it(self.tagger.load_nat)
-            except Exception as e:
-                log.error("Browser integration failed with %r on line %r", e, orig_line)
-                return False
-            log.error("Browser integration failed: cannot parse %r", orig_line)
-            return False
+class RequestHandler(BaseHTTPRequestHandler):
+
+    def do_OPTIONS(self):
+        origin = self.headers['origin']
+        if _is_valid_origin(origin):
+            self.send_response(204)
+            self.send_header('Access-Control-Allow-Origin', origin)
+            self.send_header('Access-Control-Allow-Methods', 'GET')
+            self.send_header('Access-Control-Allow-Credentials', 'false')
+            self.send_header('Access-Control-Allow-Private-Network', 'true')
+            self.send_header('Access-Control-Max-Age', 3600)
+            self.send_header('Vary', 'Origin')
+            self.end_headers()
+        else:
+            self.send_response(401)
+            self.end_headers()
 
+    def do_GET(self):
         try:
-            line = rawline.decode()
-            if parse_line(line):
-                conn.write(response(200))
+            self._handle_get()
+        except Exception as e:
+            self._response(500, str(e))
```

Comment:
While it is likely safer to catch exceptions here, I think we shouldn't return exception message in response, but rather log it.

Reply:
Fixed it. And yes, handling the exceptions here is kind of mandatory. The behavior without it is really not nice, the request would just not receive any answer